pkg/k8shandler/recovery.go

```diff
@@ -0,0 +1,354 @@
+	for nodeIndex, node := range nodesToMatch {
+
+		selector := map[string]string{
+			"cluster-name":   er.cluster.Name,
+			"es-node-client": "false",
+			"es-node-data":   "false",
+			"es-node-master": "false",
+		}
+
+		for _, role := range node.Roles {
+			switch role {
+			case loggingv1.ElasticsearchRoleClient:
+				selector["es-node-client"] = "true"
+				break
+			case loggingv1.ElasticsearchRoleData:
+				selector["es-node-data"] = "true"
+				break
+			case loggingv1.ElasticsearchRoleMaster:
+				selector["es-node-master"] = "true"
+			}
+		}
+
+		if isDataNode(node) {
+			var deploymentList *appsv1.DeploymentList
+			deploymentList, err := GetDeploymentList(er.cluster.Namespace, selector, er.client)
+			if err != nil {
+				log.Error(err, "Unable to retrieve Deployment list while recovering", "cluster", er.cluster.Name, "namespace", er.cluster.Namespace)
+				return err
+			}
+
+			for _, deployment := range deploymentList.Items {
+				clusterName, _, uuid := parseNodeName(deployment.Name)
+
+				if clusterName != er.cluster.Name {
+					continue
+				}
+
+				if sliceContainsString(knownUUIDs, uuid) {
+					log.Info("already found while adopting", "knownUUIDs", knownUUIDs, "uuid", uuid)
+					continue
+				}
+
+				if er.cluster.Spec.Nodes[nodeIndex].GenUUID == nil {
+					if er.deploymentSpecMatchNode(node, deployment.Spec, uuidCounts[uuid]) {
+						er.cluster.Spec.Nodes[nodeIndex].GenUUID = &uuid
+						knownUUIDs = append(knownUUIDs, uuid)
+
+						er.setUUID(nodeIndex, uuid)
+						break
+					}
+				}
+			}
+		} else {
+			var statefulsetList *appsv1.StatefulSetList
+			statefulsetList, err := GetStatefulSetList(er.cluster.Namespace, selector, er.client)
+			if err != nil {
+				log.Error(err, "Unable to retrieve Statefulset list while recovering", "cluster", er.cluster.Name, "namespace", er.cluster.Namespace)
+				return err
+			}
+
+			for _, statefulSet := range statefulsetList.Items {
+				clusterName, _, uuid := parseNodeName(statefulSet.Name)
+
+				if clusterName != er.cluster.Name {
+					continue
+				}
+
+				if sliceContainsString(knownUUIDs, uuid) {
+					log.Info("already found while adopting", "knownUUIDs", knownUUIDs, "uuid", uuid)
+					continue
+				}
+
+				if er.cluster.Spec.Nodes[nodeIndex].GenUUID == nil {
+					if er.statefulSetSpecMatchNode(node, statefulSet.Spec) {
+						er.cluster.Spec.Nodes[nodeIndex].GenUUID = &uuid
+						knownUUIDs = append(knownUUIDs, uuid)
+
+						er.setUUID(nodeIndex, uuid)
+						break
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// for PVCs we only need to match on roles and replicas for the sake of naming
+func (er *ElasticsearchRequest) recoverFromPVCs(knownUUIDs []string, nodesToMatch map[int]loggingv1.ElasticsearchNode) error {
+
+	selector := map[string]string{
+		"logging-cluster": er.cluster.Name,
+	}
+
+	pvcList, err := GetPVCList(er.cluster.Namespace, selector, er.client)
```

Comment:
what happens if we have more pvcs then we have es nodes to which to attach?

Reply:
nothing we can do, it'll be first come first serve. In that scenario they likely had a cluster of a particular size and recreated their CR with a smaller cluster

Reply:
im thinking of the case this issue is addressing where we have multiple pvcs and admin labels them all thinking we will grab all of them

Reply:
i think that will be a matter of documenting this correctly to call this functionality out

Reply:
Documentation is needed and is the fall back to inform users's "we told you so", but I still believe you have some defensive coding to be concerned with here.  We inevitably will have someone label all PVCs, including the original set and there exists a case where we have more PVCs then ES nodes.  I would consider doing something like:
* fetch PVCs by label
* sort list by name
* use # PVCs = # ES nodes

Additionally updating the documentation describing the behavior.  An additional scenario to think of is say an admin who thinks he might recover simply by adding more ES nodes to satisfy the number of PVCs.  I suspect this would fail because the internal cluster ID for each set would be different.

Reply:
@jcantrill how is what is currently implemented different from your suggestion? we currently only iterate over the ES nodes that are missing UUIDs, so the number of PVCs we pick up is limited by the number of ES nodes that will be created.